### DISCUSSION TITLE
How to attach metadata to a partition and not to an asset?

### DISCUSSION THREAD

**Discussion by kodaho** [dagster-io/dagster#17557] [Q&A]
Attaching metadata to an asset can be done in the `@asset` decorator, which is very useful for passing information available during the asset definition, and using it later (for example in the IO manager).

In the same way, is it possible to pass information to a specific partition instead of the asset?
One solution could be to set all the information in the partition key but this solution is unsuitable if there is a lot of metadata.

Thanks for your help!

**Comment by sryza** [comment]
Hi @kodaho - are you able to provide a little more detail on your use case? What kind of metadata do you want to include on the partition? What kind of `PartitionsDefinition` are you using? Do multiple assets share this `PartitionsDefinition`, and do you want the same metadata to apply for the same partition across multiple assets?

**Reply by kodaho** [reply]
Hi @sryza, sorry for the lack of details 😬

I am currently exploring the possibility of using partitions to handle an extract & load pipeline in a multi-tenant config. With my current architecture, the extract is materialized by an asset, the load by the IO manager, and the multi-tenancy is handled by the use of static partitions.

The kind of metadata I want to include in each partition is a set of key → value, where the keys and values are strings. Each partition will have the same keys but different values. But a same partition across different assets will have the same keys and values.

Maybe an example will be clearer:

```python
from dagster import (
    ConfigurableIOManager,
    Definitions,
    InitResourceContext,
    ResourceParam,
    StaticPartitionsDefinition,
    asset,
    define_asset_job,
    in_process_executor,
)
from dagster._config.pythonic_config import ConfigurableResourceFactory

TENANTS = [
    {"tenant": "amazon", "key": "foo", "id": "123"},
    {"tenant": "amazon", "key": "bar", "id": "456"},
    {"tenant": "google", "key": "baz", "id": "789"},
    {"tenant": "microsoft", "key": "fooz", "id": "012"},
]

partitions = StaticPartitionsDefinition(
    [f"{tenant['tenant']}-{tenant['key']}" for tenant in TENANTS]
)


class APIClient:
    def __init__(self, tenant_id: str, key: str):
        self.tenant_id = tenant_id
        self.key = key

    def request(self, endpoint: str):
        print(f"API request for {self.tenant_id}.{endpoint} with key {self.key}")


class MyAPIResource(ConfigurableResourceFactory[APIClient]):
    tenant_id: str | None = None
    key: str | None = None

    def create_resource(self, context: InitResourceContext) -> APIClient:
        assert context.dagster_run
        tenant, _, key = context.dagster_run.tags["dagster/partition"].partition("-")
        return APIClient(self.tenant_id or tenant, self.key or key)


class MyIOManager(ConfigurableIOManager):
    def load_input(self, context):
        print(f"Loading input of partition {context.partition_key}")

    def handle_output(self, context, obj):
        # >>> How to access the values `id` of each tenant? <<<
        print(f"Handling output of partition {context.partition_key}")


@asset(name="user", partitions_def=partitions)
def user(api: ResourceParam[APIClient]):
    api.request("user")


@asset(name="purchase", partitions_def=partitions)
def purchase(api: ResourceParam[APIClient]):
    api.request("purchase")


def definitions():
    assets = [user, purchase]
    job = define_asset_job(
        name="ingestion",
        selection=assets,
        executor_def=in_process_executor,
    )
    return Definitions(
        assets=assets,
        resources={
            "api": MyAPIResource.configure_at_launch(),
            "io_manager": MyIOManager(),
        },
        jobs=[job],
    )


defs = definitions()
```

A solution here would be to re-use the list `TENANTS` in the IO manager but I was wondering if there is a way to make Dagster "aware" of the metadata for each partition.

> [!NOTE]
> I tried to explore what's behind a partition to check if I could create custom partition objects storing all the metadata. But from what I understood, a partition is just defined by its key which should be a string.

I hope it's clearer.

**Reply by sryza** [reply]
Got it - that makes sense. When you say make Dagster "aware", would this metadata ideally be displayed in the UI?

There isn't currently a way to do this, but filing a Github issue would help us track this request.

**Reply by kodaho** [reply]
Thanks for the reply @sryza!

Yes ideally, the metadata will be displayed in the UI, in the same way as they are displayed at the asset level.

I will check if there is no Github issue going that way and open one if not.
